contracts/src/NativeTokenDestination.sol

```diff
@@ -0,0 +1,321 @@
+     * are transferred from the source chain into this destination chain, and divided by `tokenMultiplier` when
+     * tokens are transferred from this destination chain back to the source chain. This is intended
+     * when the "decimals" value on the source chain is less than the native EVM denomination of 18.
+     * If multiplyOnReceive is false, the raw token amount value will be divided by `tokenMultiplier` when tokens
+     * are transferred from the source chain into this destination chain, and multiplied by `tokenMultiplier` when
+     * tokens are transferred from this destination chain back to the source chain.
+     */
+    bool public immutable multiplyOnReceive;
+
+    /**
+     * @notice The wrapped native token contract that represents the native tokens on this chain.
+     */
+    IWrappedNativeToken public immutable token;
+
+    constructor(
+        address teleporterRegistryAddress,
+        address teleporterManager,
+        bytes32 sourceBlockchainID_,
+        address tokenSourceAddress_,
+        address feeTokenAddress_,
+        uint256 initialReserveImbalance_,
+        uint256 decimalsShift,
+        bool multiplyOnReceive_
+    )
+        TeleporterTokenDestination(
+            teleporterRegistryAddress,
+            teleporterManager,
+            sourceBlockchainID_,
+            tokenSourceAddress_,
+            feeTokenAddress_
+        )
+    {
+        token = IWrappedNativeToken(feeTokenAddress);
+
+        require(
+            initialReserveImbalance_ != 0, "NativeTokenDestination: zero initial reserve imbalance"
+        );
+
+        initialReserveImbalance = initialReserveImbalance_;
+        currentReserveImbalance = initialReserveImbalance_;
+
+        require(decimalsShift <= 18, "NativeTokenDestination: invalid decimalsShift");
+        tokenMultiplier = 10 ** decimalsShift;
+        multiplyOnReceive = multiplyOnReceive_;
+    }
+
+    /**
+     * @notice Receives native tokens transferred to this contract.
+     * @dev This function is called when the token bridge is withdrawing native tokens to
+     * transfer to the recipient. The caller must be the wrapped native token contract.
+     */
+    receive() external payable {
+        require(
+            msg.sender == feeTokenAddress, "NativeTokenDestination: invalid receive payable sender"
+        );
+    }
+
+    /**
+     * @dev See {INativeTokenBridge-send}.
+     */
+    function send(SendTokensInput calldata input) external payable {
+        require(
+            currentReserveImbalance == 0, "NativeTokenDestination: contract undercollateralized"
+        );
+        uint256 amount = msg.value;
+        require(
+            amount > input.primaryFee + input.secondaryFee,
+            "NativeTokenDestination: insufficient amount to cover fees"
+        );
+
+        // TODO we need to guarantee that this function deposits the whole amount, or find a workaround.
```

Comment:
Is this comment referring to the secondary fee?

I think a reasonable approach would be to always deposit the secondary fee as well, and leave it up to the user to not provide the secondary fee if it's not a multihop transfer.

Reply:
So, `_deposit` just converts native tokens to ERC20 tokens, so they can be set as teleporter fees, but the secondary fee needs be on the source chain so it can be paid to the relayer of the second message on the source chain. The only reason we deposit anything here is so we can pay the primary teleporter fee. 

The comment is referring to the fact that, _**if**_ the `_deposit` function does not credit the entire amount of `input.primaryFee` in ERC20 tokens (which I don't actually know any implementation that would do this), then the teleporter contract won't actually have the sufficient balance to pay the rewards to the relayer. I think Matt implemented a safe deposit function somewhere that I need to integrate here.